helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java

```diff
@@ -335,6 +335,42 @@ public static Set<AssignableReplica> findToBeAssignedReplicasForMinActiveReplica
     return toBeAssignedReplicas;
   }
 
+  /**
+   * Merge entries from currentResourceAssignment to newAssignment.
+   * To handle minActiveReplica for delayed rebalance, new assignment is computed based on enabled live instances, but
+   * could miss out current partition allocation still on offline instances (within delayed window).
+   * The merge process is independent for each resource; for each resource-partition, it adds the <instance, state> pair
+   * to newAssignment if it's not there yet; in other word, the entries in newAssignment won't be override.
+   * @param newAssignment newAssignment to merge, this map is getting updated during this method.
+   * @param currentResourceAssignment the current resource assignment
+   * @param enabledLiveInstances the set of enabled live instance
+   */
+  public static void mergeAssignments(Map<String, ResourceAssignment> newAssignment,
+      Map<String, ResourceAssignment> currentResourceAssignment,
+      Set<String> enabledLiveInstances) {
+    // merge with current assignment for partitions assigned on rest of the instances (not immediately live)
+    currentResourceAssignment.entrySet().parallelStream().forEach(entry -> {
+      String resourceName = entry.getKey();
+      ResourceAssignment currentAssignment = entry.getValue();
+      for (Partition partition : currentAssignment.getMappedPartitions()) {
+        currentAssignment.getReplicaMap(partition).entrySet().stream() // <instance, state>
+            // the existing partitions on the enabledLiveInstances are pre-allocated, only process for the rest
+            .filter(e -> !enabledLiveInstances.contains(e.getKey()) ||
+                !newAssignment.containsKey(resourceName) ||
+                !newAssignment.get(resourceName).getReplicaMap(partition).containsKey(e.getKey()))
```

Comment:
I forgot the last PR. When we recompute on the new assignment, was that partition map only contains the single messed up replica or entire reassigned partition.

Reply:
It's a little complicated. The computed assignment contains both existing allocation and the newly computed assignment, BUT, that only apply to the "assignable" nodes, in this case, only the immediate live nodes.
Some unit tests are using https://github.com/apache/helix/blob/master/helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/constraints/MockRebalanceAlgorithm.java that doesn't assign current allocation, and it's confusing sometime.

Reply:
I know the computed result is merging two of them. I would like to for the input of newAssignment, was just containing single replica need to be reassigned.

The reason is that this logic is very complicate. I am thinking of what is the best way to merge them.

Reply:
I see your point. I thought about similar approach too, unfortunately the current logic in ClusterModel requires we pre-allocate existing partitions, otherwise the new assignment calculations won't honor the constraints. And the current logic is to pass on existing assignment.

Reply:
Updated a bit, we can filter out just the new assignment and do merge on that only.